### ISSUE TITLE
Processors which support multiple element types do not work as expected

### ISSUE THREAD

**Issue by raghav-deepsource** [INRIA/spoon#4259]
It seems that when a processor implements multiple process methods of types other than the overriding one, spoon does not behave as intended, and ProcessingVisitor applies an improper check when deciding to process a particular node type with the affected processor. the specific code that causes this error is in the `canBeProcessed` method from ProcessingVisitor:

```java
        private boolean canBeProcessed(CtElement e) {
		if (!factory.getEnvironment().isProcessingStopped()
				&& processor.getProcessedElementTypes() != null) {
			for (Object o : processor.getProcessedElementTypes()) {
				if (!((Class<?>) o).isAssignableFrom(e.getClass())) { <------ Here
					return false;
				}
			}
			return true;
		}
		return false;
	}

```

The marked check may have been intended to perform something like: "if *any* of the types this processor accepts matches the type of the given element, return true". However, it seems that it instead does the opposite and rejects any element unless it happens to match the type of the first recorded element type for this processor. 

was this intended behavior?

**Comment by slarse** [comment]
Could you give a concrete example of "a processor [that] implements multiple process methods of types other than the overriding one"? I don't really understand what you mean by that.

**Comment by raghav-deepsource** [comment]
take something like this:

```java

class SomeProcessor extends AbstractProcessor<CtIf> {
        @Override
        void process(CtIf elem) {
            System.out.println("processed " + elem.toString());
        }

        void process(CtSwitch elem) {
            System.out.println("processed " + elem.toString());
        }
}

```

The `AbstractProcessor` class's constructor looks for any overloads of `process(T)` and adds the element types to the list of accepted types:

```java
	public AbstractProcessor() {
		for (Method m : getClass().getMethods()) {
			if ("process".equals(m.getName()) && (m.getParameterTypes().length == 1)) {
				Class<?> c = m.getParameterTypes()[0];
				if (CtElement.class != c) {
					addProcessedElementType((Class<E>) c);
				}
			}
		}
		if (processedElementTypes.isEmpty()) {
			addProcessedElementType(CtElement.class);
		}
	}
```

so this may result in our processor's `processedElementTypes` list looking like:

```java
Arrays.asList(
    CtIf.class,
    CtSwitch.class,
);
```

If we try to process a switch clause using this processor though, it will not work. The reason being, that the `canBeProcessed` method will be given a CtSwitch element. When we reach the part where the method loops over the list of allowed element types, we will reject the element on the first iteration (CtSwitch is not assignable from CtIf, or vice-versa).

**Comment by slarse** [comment]
Okay, I understand what you mean now.

I wasn't around when this part of the code base was written, but I can't imagine that a processor processing multiple unrelated types was ever meant to be a supported use case. The type parameter in `AbstractProcessor` represents the type of element that it processes, so I think it would be entirely confusing to allow for processing of other types.

@monperrus perhaps you can shed some light on the design decisions here? It kind of looks like processing of multiple types would be a feature (but it obviously doesn't work). I don't think it should be a feature, it's needlessly complicated (just create two processors?) and automatically violates the single responsibility principle.

**Comment by raghav-deepsource** [comment]
> (just create two processors?)

Not very palatable...

I have worked around the existing implementation in my own code, but wanted to draw attention to this in case anybody else also encountered this and wanted a reason (or a fix) for this behavior.

**Comment by raghav-deepsource** [comment]
I do agree that this does not really work well with how spoon's processor architecture was designed, and this is in the end mostly documented by source code alone.

**Comment by slarse** [comment]
> > (just create two processors?)

> Not very palatable...

How so?

> I have implemented a workaround for this which suits my requirements in my own code

I don't really understand how this would even work. Even if you do put multiple overloads of `process` in there, the overload to call is determined at compile time. How are you dynamically selecting the "correct" method for a given element type?

**Comment by raghav-deepsource** [comment]
in some cases, one may want to analyze multiple elements while keeping some context. I essentially just have the checker's element type be CtElement. within the overriding process method, I check the type of the element and direct it to the correct overloading method there.

**Comment by slarse** [comment]
> in some cases, one may want to analyze multiple elements while keeping some context. I essentially just have the checker's element type be CtElement. within the overriding process method, I check the type of the element and direct it to the correct overloading method there.

So you're doing something like this?

```java
        @Override
        void process(CtElement elem) {
            if (elem instanceof CtType) {
                process((CtType<?>) elem);
            } else if (elem instanceof CtIf) {
                process((CtIf) elem);
            } // etc
        }
```

That's not possible to do in Spoon itself because it requires statically knowing the available overloads, which obviously we can't. Besides, it's something of an abuse of overloads, and would be better served by simply defining unique methods to avoid having noop casts (i.e. when you cast to say `CtIf`, but there's no overload for `CtIf` so it just endlessly recurses back to the "base" method).

If you're doing it some other way, feel free to share, this was just what I imagined from your description.

To summarize, there is simply no way to support this use case in Spoon using only overloading due to how method resolution works in Java. To support the basic premise of the use case, we'd probably need to redesign the way processors are invoked. There would basically need to be a unique method for each type, e.g. `processCtType`, `processCtIf` etc, (like the scanners work, for example).

**Comment by raghav-deepsource** [comment]
Understandable. 

> So you're doing something like this?

I promise it looks much better in kotlin.

**Comment by slarse** [comment]
> > So you're doing something like this?
> 
> I promise it looks much better in kotlin.

Haha, I mean, name one thing that doesn't look better in Kotlin.

I think that's also the only way you can achieve this behavior with the current architecture. Well, you don't necessarily need overloads, but you definitely need to cast the argument to some other static type to be able to dispatch to the "correct" method. It's not ideal but with behavior that dynamic, IMO it's not _that_ bad. And as noted, smart casts in Kotlin makes it look pretty neat.

The fact that one can have multiple overloads, and Spoon actually makes an appearance of taking them into account, is however misleading.

@monperrus Can you shed some light on why it's like this? Why does `AbstractProcessor` resolve several  types for processing, when it by design can only process one type?

**Comment by monperrus** [comment]
Yes, processors (under `AbstractProcessor`) are meant to support one single element type, and we never wanted/claimed to support multiple element types.

Possible actions:

* make this clearer in the API doc (would you make a PR? Thanks!)
* add a new feature to support processors for multiple element types, which is related to roadmap element https://github.com/INRIA/spoon/issues/1075

**Comment by raghav-deepsource** [comment]
I think the way to go would be to introduce a different processor API... while it would increase the size of spoon's API, it would allow us to sidestep the current implementation's limitations

**Comment by monperrus** [comment]
> I think the way to go would be to introduce a different processor API.

yes, it's a great idea, feel free to PR.

**Comment by raghav-deepsource** [comment]
Will do.
